data_steward/vocabulary.py

```diff
@@ -1,11 +1,179 @@
+
+
+def _transform_csv(in_fp, out_fp, err_fp=None):
+    if not err_fp:
+        err_fp = sys.stderr
+    csv_reader = csv.reader(in_fp, delimiter=DELIMITER)
+    header = next(csv_reader)
+    date_indexes = []
+    for i in range(0, len(header)):
+        if header[i].endswith('_date'):
+            date_indexes.append(i)
+    csv_writer = csv.writer(out_fp, delimiter=DELIMITER, lineterminator=LINE_TERMINATOR)
+    csv_writer.writerow(header)
+    for row_index, row in enumerate(csv_reader):
+        try:
+            for i in date_indexes:
+                row[i] = format_date_str(row[i])
+            csv_writer.writerow(row)
+        except Exception, e:
+            message = 'Error %s transforming row:\n%s' % (e.message, row)
+            err_fp.write(message)
+
+
+def transform_file(file_path, out_dir):
+    """
+    Format file date fields and standardize line endings a local csv file and save result in specified directory
+
+    :param file_path: Path to the csv file
+    :param out_dir: Directory to save the transformed file
+    """
+    file_name = os.path.basename(file_path)
+    table_name, _ = os.path.splitext(file_name.lower())
+    out_file_name = os.path.join(out_dir, file_name)
+    err_dir = os.path.join(out_dir, ERRORS)
+    err_file_name = os.path.join(err_dir, file_name)
+    if not os.path.exists(err_dir):
+        os.makedirs(err_dir)
+    with open(file_path, 'rb') as in_fp, open(out_file_name, 'wb') as out_fp, open(err_file_name, 'wb') as err_fp:
+        _transform_csv(in_fp, out_fp, err_fp)
+
+
+def transform_files(in_dir, out_dir):
+    """
+    Transform vocabulary files in a directory and save result in another directory
+
+    :param in_dir: Directory containing vocabulary csv files
+    :param out_dir: Directory to save the transformed file
+    """
+    fs = os.listdir(in_dir)
+    for f in fs:
+        in_path = os.path.join(in_dir, f)
+        transform_file(in_path, out_dir)
+
+
+def get_aou_general_version():
+    return hash_dir(AOU_GENERAL_PATH)
+
+
+def get_aou_general_vocabulary_row():
+    aou_gen_version = get_aou_general_version()
+    # vocabulary_id vocabulary_name vocabulary_reference vocabulary_version vocabulary_concept_id
+    return DELIMITER.join([AOU_GEN, AOU_GEN, AOU_GEN_VOCABULARY_REFERENCE, aou_gen_version, AOU_GEN_VOCABULARY_CONCEPT_ID])
+
+
+def append_concepts(in_path, out_path):
+    with open(out_path, 'wb') as out_fp:
+        # copy original rows line by line for memory efficiency
+        with open(in_path, 'rb') as in_fp:
+            for row in in_fp:
+                if AOU_GEN in row:
+                    # skip it so it is appended below
+                    warnings.warn(ERROR_APPENDING.format(in_path=in_path))
+                else:
+                    out_fp.write(row)
+
+        # append new rows
+        with open(AOU_GENERAL_CONCEPT_CSV_PATH, 'rb') as aou_gen_fp:
+            _ = next(aou_gen_fp)  # skip header
```

Comment:
You can also skip headers in csv files with or without knowing they exist by:
```
has_header = csv.Sniffer.has_header(aou_gen_fp.read(1024))
aou_gen_fp.seek(0)
if has_header:
    next(aou_gen_fp)
```

This way allows you to not care if the header does or doesn't exist for your input file.